sycl/doc/extensions/PropertyList/SYCL_EXT_ONEAPI_property_list.asciidoc

```diff
@@ -0,0 +1,711 @@
+
+runtime property:: A property that has property parameters determined at runtime and stored as members of the property.
+
+compile-time-constant property:: A property that has no parameters that are
+determined at runtime.  This includes properties that have no parameters and
+properties where all parameters are determined at compile time.  If such a
+property has parameters, their values are stored as template arguments of the
+property value class.
+
+property value:: An object of the property value class.
+For runtime properties and properties without parameters the value type is the same as the property type.
+For compile time properties with parameters the value type is given by the `value_t` type alias of the property class.
+
+
+== Proposal
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros".  An
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_PROPERTY_LIST` to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's APIs the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version.  Base features are supported.
+|===
+
+=== Representation of a property
+
+Both runtime and compile-time-constant properties are represented as class
+declarations, and by convention, these classes are declared in the root of the
+`sycl::ext::oneapi` namespace.  The class representation of a runtime property
+has no specific requirements. A runtime property typically has a constructor which takes
+the value(s) of the properties and member function(s) which return those
+values:
+
+```c++
+namespace sycl {
+namespace ext {
+namespace oneapi {
+
+// This is a runtime property with one integer parameter.
+struct foo {
+  foo(int);
+  int value;
+};
+
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+```
+
+The class representation of a compile-time-constant property must define a type
+alias named `value_t`.  `value_t` is templated on those parameters, and it is an alias to
+an instantiation of the `property_value` class which holds the
+values of the compile-time parameters.  Note that the parameters to a compile-
+time-constant property can be either types or non-type values.
+
+```c++
+namespace sycl {
+namespace ext {
+namespace oneapi {
+
+// This property has no parameters.
+struct bar {
+  using value_t = property_value<bar>;
+};
+
+// This property has one integer non-type parameter.
+struct baz {
+  template<int K>
+  using value_t = property_value<baz, std::integral_constant<int, K> >;
```

Comment:
Why not directly
```suggestion
  using value_t = property_value<baz, K>;
```
?

Reply:
C++ doesn't (yet) support universal template template arguments and therefore we need template arguments to be always types. See also https://github.com/intel/llvm/pull/4280#discussion_r688992583

Reply:
Too bad.